### ISSUE TITLE
Consistent SIGSEGV crashes, usually after ~3 days of uptime

### ISSUE THREAD

**Issue by kqarryzada** [corretto/corretto-11#202]
### Describe the bug
Consistent segmentation faults during the processing of an application. These have mostly been related to garbage collection, but some of the most recent ones seem to be unrelated to GC.

### To Reproduce
We have not been able to reproduce this outside of our client's environment, though we do have core dump logs. I've provided some more detail below.

### Expected behavior
No segmentation faults during the application run.

### Screenshots
N/A

### Platform information
    OS: Intel(R) Xeon(R) Platinum 8259CL CPU @ 2.50GHz, 16 cores, 62G, Amazon Linux release 2 (Karoo)
    Version: OpenJDK Runtime Environment Corretto-11.0.12.7.1 (11.0.12+7) (build 11.0.12+7-LTS)

### Additional context

We are providing support for a Java product for a client. In the client's production environment, they are experiencing consistent SIGSEGV crashes of different natures, and wish to help identify the cause of these issues. Most recently, we have asked them to upgrade to the latest version of Corretto, but they have continued to see JVM crashes. I initially tried to raise this with the Adoptium project in https://github.com/adoptium/adoptium-support/issues/344, but they redirected me here in case a Corretto-specific change is related. If this is not the case, and this issue should instead be evaluated by Adoptium or another group, please let me know and I will redirect this as appropriate.

I've attached the three of the most recent hotspot error logs. These are the only crashes that have happened on the latest JDK:
[A1_25082021_hs_err_pid7694.log](https://github.com/corretto/corretto-11/files/7063026/A1_25082021_hs_err_pid7694.log)
[A1_hs_err_pid16304.log](https://github.com/corretto/corretto-11/files/7063028/A1_hs_err_pid16304.log)
[A2_hs_err_pid32230.log](https://github.com/corretto/corretto-11/files/7063029/A2_hs_err_pid32230.log)

Unfortunately, we don't have a lot of detail on reproducing this crash, since we have not been able to do so outside of this environment. Most of the crashes in the past have been during garbage collection, but the most recent ones happen in compiled frames. I do have core dumps for these crashes, though. I cannot share them in their entirety because they contain PII, but I can share any necessary information privately.

Again, if this is not the correct place to receive assistance for this, please let me know. Thank you.

**Comment by eastig** [comment]
Hi @kqarryzada,

I've browsed the error reports you provided. They have different crashes. In JDK Bug System I found https://bugs.openjdk.java.net/browse/JDK-8272405 which describes a crash in `SharedRuntime::generate_class_cast_message` like in A1_hs_err_pid16304.log.
Do you have the crashes if you use other jdk-11.0.12+7 builds, for example AdoptOpenJDK build https://adoptopenjdk.net/?
I am asking this to understand if it is Corretto specific or exists in OpenJDK 11.

**Comment by simonis** [comment]
There's also [JDK-8258982](https://bugs.openjdk.java.net/browse/JDK-8258982) which describes a crash with almost identical command line but with OpenJDK 11.0.5 so I'm pretty sure this isn't Corretto related.

The only suspect thing I can see is `-XX:RefDiscoveryPolicy=1` which seems a more esoteric, not so widely used option. The option is quite old and I don't think it is very well tested nowadays, so maybe it can lead to problems with G1?

Can you please try to remove   `-XX:RefDiscoveryPolicy=1` or explicitely set it to  `-XX:RefDiscoveryPolicy=0` (which is the default) and see if the crashes are still happening?

**Comment by kqarryzada** [comment]
Thank you both for the replies. I believe that this is an issue in OpenJDK 11 and is not Corretto-specific, since this environment also ran into a similar issue on AdoptOpenJDK before it was switched to Corretto.

@simonis, I've suggested this, and will report back if it provides any improvements.

**Comment by simonis** [comment]
@kqarryzada thanks for the confirmation that the error also happens with AdoptOpenJDK.

Please report back if you have the results from testing with `-XX:RefDiscoveryPolicy=0`. Once we know that's the reason for the failure, it might be easier to fix :)

**Comment by marcomoscatiello** [comment]
hi guys, we are having the same issues on our jenkins builds. The issue seems to be very random and we cannot really figure out what's wrong.
We run android builds on Amazon Linux 2 , Java 11 corretto , gradle 7.2 .

`15:24:56 #
15:24:56 # A fatal error has been detected by the Java Runtime Environment:
15:24:56 #
15:24:56 #  SIGSEGV (0xb) at pc=0x00007f7871be2350, pid=8082, tid=8099
15:24:56 #
15:24:56 # JRE version: OpenJDK Runtime Environment Corretto-11.0.12.7.1 (11.0.12+7) (build 11.0.12+7-LTS)
15:24:56 # Java VM: OpenJDK 64-Bit Server VM Corretto-11.0.12.7.1 (11.0.12+7-LTS, mixed mode, tiered, compressed oops, g1 gc, linux-amd64)
15:24:56 # Problematic frame:
15:24:56 # V  [libjvm.so+0xbe1350][thread 8117 also had an error]
15:24:56   Node::uncast() const+0x0
15:24:56 #
15:24:56 # No core dump will be written. Core dumps have been disabled. To enable core dumping, try "ulimit -c unlimited" before starting Java again
`
I do also have a hs_err_pid7907.log file downloaded after the job failed (i can provide it if it can help).

Any idea what this can be ?

**Comment by simonis** [comment]
@marcomoscatiello can you please attach the `hs_err_pid7907.log` file so we can compare it with the previous crashes?

Also, do you know if you're using `-XX:RefDiscoveryPolicy=0` like the previous reporter?

Thanks for reporting,
Volker

**Comment by marcomoscatiello** [comment]
@simonis , here attached the log file.
No, we are not using the DiscoveryPolicy setting.
[hs_err_pid7907.log](https://github.com/corretto/corretto-11/files/7411556/hs_err_pid7907.log)

**Comment by simonis** [comment]
Thanks for the error log file. From a quick look, it doesn't seem to be related to the this issues. It's a crash in the C2 compiler. It is however almost identical to [JDK-8206441](https://bugs.openjdk.java.net/browse/JDK-8206441) (see the corresponding [hs_error.log](https://bugs.openjdk.java.net/secure/attachment/77519/hs_error.log) file) and [JDK-8181499](https://bugs.openjdk.java.net/browse/JDK-8181499).

I've uploaded your error log file together with some comments to  [JDK-8181499](https://bugs.openjdk.java.net/browse/JDK-8181499).

Some *interesting* things I noticed when looking at your error log file are:
- you diable verification (is this required?, can you try with verification?)
- your `JAVA_HOME` environment variable seems to point to a jdk8 installation (`/home/ec2-user/.jenv/versions/1.8.0.252`) although you are running jdk11. Are you aware of this?

**Comment by marcomoscatiello** [comment]
hi @simonis , thank you very much for pointing that out!
So i fixed that JAVA_HOME env var that was set by jenv (a tool to switch java versions) and now JAVA_HOME env var is also pointing to the amazon java11 corretto .
the issue is that even after fixing it, we still get the segmentation fault.
I'm uploading the log file here (the new one)
I hope this can help. thank you very much

[hs_err_pid18919.log](https://github.com/corretto/corretto-11/files/7418514/hs_err_pid18919.log)
.

**Comment by simonis** [comment]
@marcomoscatiello I didn't really expected that fixing JAVA_HOME will solve the problem :(

But  thanks for trying and thanks for submitting the new error file. I looks like the problem is quite reproducible on your side. The new error file has exactly the same stack trace like the previous one. It also crashed at the same time (i.e. ~9min after start).

Do you think you can create a smaller test case which we can reproduce on our side (e.g. by looking into the gradle log files to see which exact test caused the crash)? That would be really helpful for us. From the `hs_err.log` file I can only see that you were executing `worker.org.gradle.process.internal.worker.GradleWorkerMain 'Gradle Test Executor 1'`.

Alternatively, if that's not possible at all, could you re-run your tests with a debug version of Corretto. The debug version has a lot of additional checks and assertions which might help to pinpoint the root cause of the problem before the crash happens. I can send you such a version if you want to try it out.

And just for completeness, have you also tried to remove the `-Xverify:none` flag (or replace it with `-Xverify:all`). From the error log I can see that the application is dynamically generating bytecode. I just want to make sure that this bytecode is correct.

**Comment by kqarryzada** [comment]
@simonis, thanks for your help and patience. We asked our client to remove `-XX:RefDiscoveryPolicy=1` from their java arguments, and I believe this was done about three weeks ago. They have not reported a crash since. It seems like our use of this argument was very closely related to this issue. Thanks again for looking into this, and pointing that out.

**Comment by simonis** [comment]
Thanks for letting us know @kqarryzada and happy that we could help :)

**Comment by simonis** [comment]
@marcomoscatiello, @navyxliu has closed  this issue because the initial issue was fixed by removing the `-XX:RefDiscoveryPolicy=1`. As I wrote before, your crash reports are in the C2 compiler and have a different reason.

Could you please be so kind to open a new issue with your error log files and reference this issue for additional information. Did you manage to create a simpler, standalone reproducer for your issue?

Thank you and best regards,
Volker
